mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala

```diff
@@ -89,12 +106,13 @@ final class QuantileDiscretizer(override val uid: String)
   }
 
   override def fit(dataset: DataFrame): Bucketizer = {
-    val samples = QuantileDiscretizer
-      .getSampledInput(dataset.select($(inputCol)), $(numBuckets), $(seed))
-      .map { case Row(feature: Double) => feature }
-    val candidates = QuantileDiscretizer.findSplitCandidates(samples, $(numBuckets) - 1)
-    val splits = QuantileDiscretizer.getSplits(candidates)
-    val bucketizer = new Bucketizer(uid).setSplits(splits)
+    val splits = dataset.stat.approxQuantile($(inputCol),
+      (0.0 to 1.0 by 1.0/$(numBuckets)).toArray, getRelativeError)
+
+    // adjust splits to be consistent with class description
+    val adjustedSplits = Double.NegativeInfinity +:
+      splits.drop(1).dropRight(1) :+ Double.PositiveInfinity
```

Comment:
~~~scala
splits(0) =  Double.NegativeInfinity
splits(splits.length - 1) = Double.PositiveInfinity
~~~

to avoid creating temp arrays.